Libraries/src/Amazon.Lambda.RuntimeSupport/Helpers/Logging/AbstractLogMessageFormatter.cs

```diff
@@ -0,0 +1,279 @@
+            {
+                var c = messageTemplate[i];
+                switch (c)
+                {
+                    case '{':
+                        if (state == LogFormatParserState.InMessage)
+                        {
+                            state = LogFormatParserState.PossibleParameterOpen;
+                        }
+                        else if (state == LogFormatParserState.PossibleParameterOpen)
+                        {
+                            // this is an escaped brace
+                            state = LogFormatParserState.InMessage;
+                        }
+                        break;
+                    case '}':
+                        if (state != LogFormatParserState.InMessage)
+                        {
+                            if(paramStartIdx != -1)
+                            {
+                                // Since we have a closing bracket and there is at least a start to a message property label 
+                                // then we know we have hit the end of the message property.
+                                messageProperties.Add(new MessageProperty(messageTemplate.AsSpan().Slice(paramStartIdx, i - paramStartIdx)));
+                            }
+
+                            state = LogFormatParserState.InMessage;
+                            paramStartIdx = -1;
+                        }
+                        break;
+                    default:
+                        if (state == LogFormatParserState.PossibleParameterOpen)
+                        {
+                            // non-brace character after '{', transition to InParameter
+                            paramStartIdx = i;
+                            state = LogFormatParserState.InParameter;
+                        }
+                        break;
+                }
+            }
+
+            var readonlyMessagesProperties = messageProperties.AsReadOnly();
+
+            // If there is a room in the message template cache then cache the parse results for
+            // later logging performance increase.
+            if (MESSAGE_TEMPLATE_PARSE_CACHE.Count < MESSAGE_TEMPLATE_PARSE_CACHE_MAXSIZE)
+            {
+                MESSAGE_TEMPLATE_PARSE_CACHE.TryAdd(messageTemplate, readonlyMessagesProperties);
+            }
+
+            return readonlyMessagesProperties;
+        }
+
+        /// <summary>
+        /// Subclasses to implement to format the message given the requirements of the subclass.
+        /// </summary>
+        /// <param name="state"></param>
+        /// <returns></returns>
+        public abstract string FormatMessage(MessageState state);
+
+        internal const string DateFormat = "yyyy-MM-ddTHH:mm:ss.fffZ";
+
+        /// <summary>
+        /// Format the timestamp of the log message in format Lambda service prefers.
+        /// </summary>
+        /// <param name="state"></param>
+        /// <returns></returns>
```

Comment:
nit: add docs for the empty tags in this class

Reply:
Done